servicetalk-concurrent-api/src/main/java/io/servicetalk/concurrent/api/Publisher.java

```diff
@@ -137,7 +138,57 @@ protected Publisher() {
      * @see <a href="http://reactivex.io/documentation/operators/filter.html">ReactiveX filter operator.</a>
      */
     public final Publisher<T> filter(Predicate<? super T> predicate) {
-        return new FilterPublisher<>(this, predicate);
+        return filter(() -> predicate);
+    }
+
+    /**
+     * Filters items emitted by this {@link Publisher}.
+     * <p>
+     * This method provides a data transformation in sequential programming similar to:
+     * <pre>{@code
+     *     List<T> results = ...;
+     *     Predicate<K> filter = filterSupplier.get();
+     *     for (T t : resultOfThisPublisher()) {
+     *         if (filter.test(t)) {
+     *             results.add(t);
+     *         }
+     *     }
+     *     return results;
+     * }</pre>
+     *
+     * @param filterSupplier Used to obtain a new {@link Predicate} such that if {@link Predicate#test(Object)} returns
+     * {@code true} the item is emitted to the returned {@link Publisher}. This {@link Predicate} is unique per
+     * subscribe and may therefore keep state and can be used for example to track distinct signals.
+     * @return A {@link Publisher} emits filtered signals observed by this {@link Publisher}.
+     * @see <a href="http://reactivex.io/documentation/operators/filter.html">ReactiveX filter operator.</a>
+     * @see <a href="http://reactivex.io/documentation/operators/distinct.html">ReactiveX distinct operator.</a>
+     */
+    public final Publisher<T> filter(Supplier<? extends Predicate<? super T>> filterSupplier) {
```

Comment:
With this approach, number of operators can easily double. People will come and ask for the same for map/flatMap/when*/etc. For the public API, can we just rely on users to use `defer` when they need a state per subscribe? We can keep this approach as pkg-private for use-cases like `distinct`.

Reply:
Sure this method can be made private for now. The filter API in this PR provides a simplification of `distinct` specializations in rxjava [1] and [reactor](https://github.com/reactor/reactor-core/blob/main/reactor-core/src/main/java/reactor/core/publisher/Flux.java#L4404-L4408). It is true that folks may ask for similar `Supplier` style APIs on other methods, but in this case there is precedence for it in other libraries.
[1]
```java
public final <@NonNull K> Flowable<T> distinct(@NonNull Function<? super T, K> keySelector,
            @NonNull Supplier<? extends Collection<? super K>> collectionSupplier)
```